win_shortcut: Create, manage, remove Windows shortcuts

dagwieers opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
win_shortcut

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
v2.3

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->
This modules manages Windows shortcuts and all its properties.
It is idempotent and supports check-mode.

This relates to #19694

dagwieers committed:
win_shortcut: Create, manage, remove Windows shortcuts

This modules manages Windows shortcuts and all its properties.

The module is idempotent and supports check-mode.

This relates to #19694

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    # TODO: Find a better way to do has_attr() or isinstance() ?
+    If (Get-Member -InputObject $ShortCut -Name Arguments) {
```
For COM/IDispatch objects like this without a managed interface wrapper, this is unfortunately about the best you can do (otherwise you'd just do `-is [type]`). The "technically proper" thing would be a `Marshal.QueryInterface` for the IID of WshShortcut/WshUrlShortcut, but that's pretty fugly. Given how ancient this stuff is and how unlikely it is that `Arguments` will magically appear on the WshUrlShortcut down the line, this is probably the best balance of readability/pedantic correctness. A comment explaining why you're doing it for the next guy might be good though, eg, "determine if we have a WshShortcut or WshUrlShortcut by checking for presence of Arguments property"...

dagwieers replied:
Well, the thing is (and I only learned it yesterday !) that a URL shortcut object only has a location and a URL. That is why I had to move everything out of the way if it was an URL shortcut. It looks a bit weird/wrong, but a URL shortcut is just very limited (and decided based on the extension...).

Which means I can also look at the destination to know what object it will be. But I prefer to leave that up to Windows (who knows what they will come up with in the future ;-)).

So I read from you that I need to document this more explicitly in the code, will do.
(For me this is new, but maybe this was common knowledge for Windows admins ;-))

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -0,0 +1,147 @@
+
+    # TODO: Find a better way to do has_attr() or isinstance() ?
+    If (Get-Member -InputObject $ShortCut -Name Arguments) {
+
```
Also, doesn't look like you did anything with the WshUrlShortcut case (eg, delete/re-create if necessary)

dagwieers replied:
Exactly, for a WshUrlShortcut object only the TargetPath can be managed. I will have to make that clear in the documentation. But I wonder if a hard failure is in order if it is a URL shortcut and other parameters are provided ? Currently I ignore them if it is a WshUrlShortcut.

PS I think this is pretty limited anyway, what I did at the elementary school is encapsulated the URL in a call to Chrome so that you can have a proper icon, description, etc... (I now added an example for a wshShortcut and a WshUrlShortcut for a URL, the first is more comprehensive and powerful)

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -0,0 +1,147 @@
+$icon = [System.Environment]::ExpandEnvironmentVariables($orig_icon)
+$description = [System.Environment]::ExpandEnvironmentVariables($orig_description)
+
+$result = New-Object PSObject @{
```
Not that you need to change this, but you know you can just use a raw dictionary for the result object, right? Gets rid of all the Get-Attr/Set-Attr jazz and just lets you do `$mydict.SomeKey = "blar"`

dagwieers replied:
What would you prefer ? And what would be more efficient internally ?

dagwieers replied:
Still interested in your view on this. Because I did wonder why it was a PSObject (whatever that is).
I prefer basic datatypes if possible, so I would like to rewrite it myself, but other modules do this.

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -0,0 +1,147 @@
+            Set-Attr $result "directory" $ShortCut.WorkingDirectory
+        }
+
+        If ($hotkey -ne $null -and $ShortCut.Hotkey -ne $hotkey) {
```
Again a pedantic thing, but shouldn't it be possible to set most of these to null/empty?

dagwieers replied:
I did this on purpose actually. The reason is that I want to check if it set to $null (from the parameter handling) and not explicitly set to "". Two reasons:
1. Maybe the user wants it empty
2. When environment variables are substituted, $null becomes "", so I loose whether it was originally undefined, or explicitely set to an empty string. That is why I eventually moved to $orig_ variables, and I prefer to keep the exact $null match.

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -0,0 +1,147 @@
+            Set-Attr $result "changed" $true
+            Set-Attr $result "args" $args
+            $ShortCut.Arguments = $args
+        } Else {
```
Can't you just ditch all these Else's and just always set the proper key to the value from `$Shortcut`?

dagwieers replied:
Yes, there might be something more efficient. Originally I only returned the values that were modified, but I figured that's not good practice. Will do.

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -0,0 +1,147 @@
+
+$orig_src = Get-AnsibleParam -obj $params -name "src" -default $null
+$orig_dest = Get-AnsibleParam -obj $params -name "dest" -failifempty $true
+$state = Get-AnsibleParam -obj $params -name "state" -default "present"
```
You can add `-ValidateSet "present","absent"` here and ditch the outer validation of `state`

dagwieers replied:
Alright, will do !

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.py:
```diff
@@ -0,0 +1,108 @@
+  src:
+    description:
+    - Executable or URL the shortcut points to.
+    required: false
```
Again, not something you need to change, but do you know that required: false/default: null are the defaults and are no longer required?

dagwieers replied:
I did not know, and was wondering about that :-) Thanks !

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.py:
```diff
@@ -0,0 +1,108 @@
+    - Key combination for the shortcut.
+    required: false
+    default: null
+  windowstyle:
```
Should arguably be `window_style`...

dagwieers replied:
Right, I am not sure about naming conventions. In this case I preferred windowstyle over window_style, but I cannot tell you why exactly.

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.py:
```diff
@@ -0,0 +1,108 @@
+  windowstyle:
+    description:
+    - Influences how the application is displayed when it is launched.
+        C(1) for default location and size
```
Minor UI thing, but might be nice if "default", "maximized" and "minimized" were accepted instead of the numeric enum values. Should probably also have "choices" defined (and -ValidateSet up above).

dagwieers replied:
Yes, I didn't know how to do that in Powershell, so I left it at that.
Will fix !

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.py:
```diff
@@ -0,0 +1,108 @@
+'''
+
+RETURN = '''
+'''
```
Some stuff complains if the .py file is missing the final LF- can't remember if we hit that anymore or not, but not a bad idea to add it just in case.

dagwieers replied:
I think vim did whatever was done ;-)

nitzmahone reviewed (changes requested):
Couple little things and a couple suggestions, but mostly LGTM.

jhawkesworth commented:
This looks good to me, I have looked through it and can't think of any further comments to add.

dagwieers committed:
Changes required after @nitzmahone review

jhawkesworth commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -0,0 +1,140 @@
+        }
+        # Report removal success
+        Set-Attr $result "changed" $true
+    } Else {
```
Not important but you could loose this Else { }

dagwieers replied:
Right, but I personally liked the fact that the Else is intentionally left blank (and not an oversight).

nitzmahone commented on lib/ansible/modules/windows/win_shortcut.py:
```diff
@@ -0,0 +1,106 @@
+- 'Windows has two types of shortcuts: Application and URL shortcuts. URL shortcuts only consists of C(dest) and C(src)'
+'''
+
+EXAMPLES = '''
```
This needs to be a raw string (`r'''`) to fix the failing doc sanity check- otherwise it tries to interpret `C:\Users` as a Unicode hex literal. Ah, the joys of nesting escapes across multiple languages... :)

dagwieers replied:
I thought it was complaining about the empty RETURN statement, I would never have figured out it was the section before that was the culprit :-) 

Fixed.

dagwieers committed:
Added -type "path" to parameter definitions

dagwieers commented:
So I added an option `-type "path"` to the `Get-AnsibleParam` function so that we can have it do things like expand environment variables. Now, I also added a function `Expand-Environment()` because some modules may want to expand environment variables for things that are not paths (like a description). Which is also used in win_shortcut as well.

dagwieers committed:
Small fixes

- Add conversion from window style name to window style id
- Fix error message output (Why didn't the original work ?)
